unit/util/simplify_expr.cpp

```diff
@@ -61,22 +61,67 @@ TEST_CASE("Simplify const pointer offset")
   REQUIRE(offset_value==1234);
 }
 
-TEST_CASE("Simplify Java boolean -> int -> boolean casts")
+namespace
+{
+
+void test_unnecessary_cast(const typet &type)
 {
   config.set_arch("none");
 
-  const exprt simplified=simplify_expr(
-    typecast_exprt(
+  WHEN("The casts can be removed, they are")
+  {
+    const exprt simplified=simplify_expr(
       typecast_exprt(
-        symbol_exprt(
-          "foo",
-          java_boolean_type()),
-        java_int_type()),
-      java_boolean_type()),
-    namespacet(symbol_tablet()));
-
-  REQUIRE(simplified.id()==ID_symbol);
-  REQUIRE(simplified.type()==java_boolean_type());
-  const auto &symbol=to_symbol_expr(simplified);
-  REQUIRE(symbol.get_identifier()=="foo");
+        typecast_exprt(symbol_exprt("foo", type), java_int_type()),
+        type),
+      namespacet(symbol_tablet()));
+
+    REQUIRE(simplified.id()==ID_symbol);
+    REQUIRE(simplified.type()==type);
+    const auto &symbol=to_symbol_expr(simplified);
+    REQUIRE(symbol.get_identifier()=="foo");
+  }
+
+  WHEN("Casts should remain, they are left untouched")
+  {
+    {
+      const exprt simplified=simplify_expr(
+        typecast_exprt(symbol_exprt("foo", type), java_int_type()),
+        namespacet(symbol_tablet()));
+
+      REQUIRE(simplified.id()==ID_typecast);
+      REQUIRE(simplified.type()==java_int_type());
+    }
+
+    {
+      const exprt simplified=simplify_expr(
+        typecast_exprt(symbol_exprt("foo", java_int_type()), type),
+        namespacet(symbol_tablet()));
+
+      REQUIRE(simplified.id()==ID_typecast);
+      REQUIRE(simplified.type()==type);
+    }
+  }
```

Comment:
You might consider a test that involves much more nesting `(int)(short)(int)(short)` for example

Reply:
Done.